processing/src/main/java/org/apache/druid/query/planning/DataSourceAnalysis.java

```diff
@@ -110,17 +111,19 @@ public DataSource getBaseDataSource()
   }
 
   /**
-   * If {@link #getBaseDataSource()} is a {@link TableDataSource}, returns it. Otherwise, returns an empty Optional.
+   * Unwraps the {@link #getBaseDataSource()} if its a {@link TableDataSource}.
    *
-   * Note that this can return empty even if {@link #isConcreteAndTableBased()} is true. This happens if the base
+   * @throws An error of type {@link DruidException.Category#DEFENSIVE} if the {@link BaseDataSource} is not a table.
+   *
+   * note that this may not be true even {@link #isConcreteAndTableBased()} is true - in cases when the base
    * datasource is a {@link UnionDataSource} of {@link TableDataSource}.
    */
-  public Optional<TableDataSource> getBaseTableDataSource()
+  public TableDataSource getBaseTableDataSource()
```

Comment:
Maybe we should also add a method `hasBaseTableDatasource()`, otherwise some callers need to do an `instanceof` check.

Reply:
I've removed the `instanceof` from `Queries` ; as I think that was aiming as well on an exception which could be the one this method produces as well.